Sources/PackageLoading/ManifestLoader.swift

```diff
@@ -421,6 +429,64 @@ public final class ManifestLoader: ManifestLoaderProtocol {
         }
     }
 
+    private struct Imports: Decodable {
+        let imports: [String]
+    }
+
+    private func validateImports(
+        manifestPath: AbsolutePath,
+        toolsVersion: ToolsVersion,
+        callbackQueue: DispatchQueue,
+        completion: @escaping (Result<Bool, Error>) -> Void) {
+            // If there are no import restrictions, we do not need to validate.
+            guard let restrictImports = restrictImports, toolsVersion >= restrictImports.startingToolsVersion else {
+                return callbackQueue.async {
+                    completion(.success(true))
+                }
+            }
+
+            // Allowed are the expected defaults, plus anything allowed by the configured restrictions.
+            let allowedImports = ["PackageDescription", "Swift", "SwiftOnoneSupport"] + restrictImports.allowedImports
+
+            // wrap the completion to free concurrency control semaphore
+            let completion: (Result<Bool, Error>) -> Void = { result in
+                self.concurrencySemaphore.signal()
+                completion(result)
+            }
+
+            // we must not block the calling thread (for concurrency control) so nesting this in a queue
+            self.evaluationQueue.addOperation {
+                do {
+                    // park the evaluation thread based on the max concurrency allowed
+                    self.concurrencySemaphore.wait()
+
+                    let cmd = [self.toolchain.swiftCompilerPathForManifests.pathString,
+                               manifestPath.pathString, "-scan-dependencies", "-Xfrontend", "-import-prescan"] + self.extraManifestFlags
+
+                    TSCBasic.Process.popen(arguments: cmd, environment: self.toolchain.swiftCompilerEnvironment, queue: callbackQueue) { result in
```

Comment:
no way to do this in-memory?

Reply:
not really, we would need to link the swift-driver to be able to do that and that is something we're currently avoiding for anything but the `Build` module and CLI commands.

Reply:
this means a performance hit for non-cached manifests. probably okay as it is selectively enabled?

Reply:
Yah, that was my thinking. If we use it more widely, we may need to reconsider, but since it is kind of a "trial" right now, I think this is OK.

Somewhat related question: should the restrictions become part of the cache key?

Reply:
I dont think they need to be, over time the restriction will be applied

Reply:
I think I'll refactor this slightly and extract just the import-scanning part so that it'll be easier to replace in the future (and also to re-use for other features).